<a href="https://colab.research.google.com/github/RPGraciotti/BootCampAlura/blob/main/Modulo_6/Aula_1_Contextualiza%C3%A7%C3%A3o_do_mercado_financeiro.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sessão 1 - Contexto financeiro
Introdução:

Pilares da cadeia de crédito:

1. Prospecção
2. Concessão - Foco do módulo do BootCamp
3. Gestão de Risco 
4. Recuperação

Pedido de crédito:
Instituição financeira utiliza um modelo de decisão + política de crédito

Bytebank é o banco fictício que a Alura utiliza para ferramentas didáticas e será trabalhado durante o curso.
Streamlit é uma biblioteca python que é utilizada para hospedar o modelo e gera uma interface sem que seja necessário conhecimento de front-end.

Projeto final: Simulador de avaliação de crédito utilizando o streamlit.

Objetivo é derivado de um desafio do kaggle, traduzido e manipulado para utilização: https://www.kaggle.com/rikdifos/credit-card-approval-prediction

# Sessão 2 - Primeiros passos com a base

In [1]:
import pandas as pd

In [2]:
df_cadastrados = pd.read_csv("https://github.com/RPGraciotti/BootCampAlura/raw/main/Data/clientes_cadastrados.txt")
df_cadastrados

ID_Cliente Genero  Tem_Carro  Tem_Casa_Propria  Qtd_Filhos  \
0          5008804      H          1                 1           0   
1          5008805      H          1                 1           0   
2          5008806      H          1                 1           0   
3          5008808      M          0                 1           0   
4          5008809      M          0                 1           0   
...            ...    ...        ...               ...         ...   
438552     6840104      H          0                 1           0   
438553     6840222      M          0                 0           0   
438554     6841878      M          0                 0           0   
438555     6842765      M          0                 1           0   
438556     6842885      M          0                 1           0   

        Rendimento_Anual   Categoria_de_renda Grau_Escolaridade  \
0               427500.0            Empregado   Ensino superior   
1               427500.0            Empregado   Ensino superior   
2               112500.0            Empregado      Ensino medio   
3               270000.0  Associado comercial      Ensino medio   
4               270000.0  Associado comercial      Ensino medio   
...                  ...                  ...               ...   
438552          135000.0          Pensionista      Ensino medio   
438553          103500.0            Empregado      Ensino medio   
438554           54000.0  Associado comercial   Ensino superior   
438555           72000.0          Pensionista      Ensino medio   
438556          121500.0            Empregado      Ensino medio   

         Estado_Civil           Moradia  Idade  Tem_Celular  \
0       Uniao estavel           Alugado   33.0            1   
1       Uniao estavel           Alugado   33.0            1   
2              Casado      Casa propria   59.0            1   
3            Solteiro      Casa propria   52.0            1   
4            Solteiro      Casa propria   52.0            1   
...               ...               ...    ...          ...   
438552       Separado      Casa propria   62.0            1   
438553       Solteiro      Casa propria   44.0            1   
438554       Solteiro  Mora com os pais   22.0            1   
438555         Casado      Casa propria   59.0            1   
438556         Casado      Casa propria   52.0            1   

        Tem_telefone_trabalho  Tem_telefone_fixo  Tem_email          Ocupacao  \
0                           1                  0          0            Outros   
1                           1                  0          0            Outros   
2                           0                  0          0         Seguranca   
3                           0                  1          1             Venda   
4                           0                  1          1             Venda   
...                       ...                ...        ...               ...   
438552                      0                  0          0            Outros   
438553                      0                  0          0  Construcao Civil   
438554                      1                  0          0             Venda   
438555                      0                  0          0            Outros   
438556                      0                  1          0             Venda   

        Tamanho_Familia  Anos_empregado  
0                   2.0            12.4  
1                   2.0            12.4  
2                   2.0             3.1  
3                   1.0             8.4  
4                   1.0             8.4  
...                 ...             ...  
438552              1.0         -1000.7  
438553              1.0             8.2  
438554              1.0             1.0  
438555              2.0         -1000.7  
438556              2.0             3.3  

[438557 rows x 18 columns]

In [3]:
df_cadastrados.shape
# numero de linhas, numero de colunas

(438557, 18)

In [4]:
df_cadastrados.head()

ID_Cliente Genero  Tem_Carro  Tem_Casa_Propria  Qtd_Filhos  \
0     5008804      H          1                 1           0   
1     5008805      H          1                 1           0   
2     5008806      H          1                 1           0   
3     5008808      M          0                 1           0   
4     5008809      M          0                 1           0   

   Rendimento_Anual   Categoria_de_renda Grau_Escolaridade   Estado_Civil  \
0          427500.0            Empregado   Ensino superior  Uniao estavel   
1          427500.0            Empregado   Ensino superior  Uniao estavel   
2          112500.0            Empregado      Ensino medio         Casado   
3          270000.0  Associado comercial      Ensino medio       Solteiro   
4          270000.0  Associado comercial      Ensino medio       Solteiro   

        Moradia  Idade  Tem_Celular  Tem_telefone_trabalho  Tem_telefone_fixo  \
0       Alugado   33.0            1                      1                  0   
1       Alugado   33.0            1                      1                  0   
2  Casa propria   59.0            1                      0                  0   
3  Casa propria   52.0            1                      0                  1   
4  Casa propria   52.0            1                      0                  1   

   Tem_email   Ocupacao  Tamanho_Familia  Anos_empregado  
0          0     Outros              2.0            12.4  
1          0     Outros              2.0            12.4  
2          0  Seguranca              2.0             3.1  
3          1      Venda              1.0             8.4  
4          1      Venda              1.0             8.4

In [5]:
df_cadastrados.loc[:,"ID_Cliente"].value_counts().unique() # significa que só tem ids que aparecem 1 ou 2 vezes, ou seja, só tem com no máximo uma repetição
# tem menos ids do que linhas no dataframe, indicando que podemos ter ids repetidos

array([2, 1])

In [6]:
print(df_cadastrados.duplicated().sum()) # indica que existem 0 linhas totalmente duplicadas, ou seja, os erros são nas colunas
print(df_cadastrados.loc[:,"ID_Cliente"].duplicated(keep = False).sum()) # 94 ids que estão repetidos de alguma forma

0
94


In [7]:
df_cadastrados[df_cadastrados.loc[:,"ID_Cliente"].duplicated(keep = False)] # keep = True mantém só o primeiro id repetido, 
# usando keep = True mantemos todos os valores repetidos

ID_Cliente Genero  Tem_Carro  Tem_Casa_Propria  Qtd_Filhos  \
421211     7702516      M          0                 1           2   
421268     7602432      H          0                 1           0   
421349     7602432      M          0                 0           0   
421464     7836971      H          1                 0           1   
421698     7213374      H          1                 0           0   
...            ...    ...        ...               ...         ...   
433158     7282535      M          0                 1           0   
433159     7742853      H          0                 1           0   
433217     7135270      M          0                 1           0   
433666     7091721      M          1                 1           0   
433789     7618285      M          0                 1           0   

        Rendimento_Anual   Categoria_de_renda Grau_Escolaridade  \
421211          180000.0            Empregado      Ensino medio   
421268          315000.0  Associado comercial   Ensino superior   
421349          117000.0          Pensionista   Ensino superior   
421464          157500.0            Empregado      Ensino medio   
421698          148500.0            Empregado      Ensino medio   
...                  ...                  ...               ...   
433158           63000.0          Pensionista      Ensino medio   
433159          157500.0            Empregado      Ensino medio   
433217          216000.0          Pensionista      Ensino medio   
433666           90000.0  Associado comercial      Ensino medio   
433789          157500.0            Empregado      Ensino medio   

         Estado_Civil       Moradia  Idade  Tem_Celular  \
421211         Casado  Casa propria   32.0            1   
421268  Uniao estavel  Casa propria   46.0            1   
421349         Casado  Casa propria   68.0            1   
421464         Casado  Casa propria   38.0            1   
421698         Casado  Casa propria   27.0            1   
...               ...           ...    ...          ...   
433158         Casado  Casa propria   58.0            1   
433159       Solteiro  Casa propria   41.0            1   
433217         Casado  Casa propria   63.0            1   
433666         Casado  Casa propria   39.0            1   
433789       Solteiro  Casa propria   28.0            1   

        Tem_telefone_trabalho  Tem_telefone_fixo  Tem_email          Ocupacao  \
421211                      1                  1          0             Venda   
421268                      0                  1          0         Motorista   
421349                      0                  0          0            Outros   
421464                      0                  0          0            Outros   
421698                      0                  1          0  Construcao Civil   
...                       ...                ...        ...               ...   
433158                      0                  1          0            Outros   
433159                      0                  0          0  Construcao Civil   
433217                      0                  0          0            Outros   
433666                      0                  0          0          Medicina   
433789                      0                  0          0  Construcao Civil   

        Tamanho_Familia  Anos_empregado  
421211              4.0             3.4  
421268              2.0             3.6  
421349              2.0         -1000.7  
421464              3.0            15.1  
421698              2.0             2.6  
...                 ...             ...  
433158              2.0         -1000.7  
433159              1.0             4.6  
433217              2.0         -1000.7  
433666              2.0             6.2  
433789              1.0             2.8  

[94 rows x 18 columns]

# Desafio 1: Explorar IDs repetidos (Atualização de cadastro? Erro?)

Exemplo do cliente 7602432: Tem duas entradas, a segunda entrada tem valores muito discrepantes, levando a crer ser um erro. Mas o mais interessante é a coluna "Anos_empregado" dando um valor de -1000. Isso certamente parece um erro de preenchimento. Atentar para verificar se isso se repete.

In [8]:
ids_repetidos = df_cadastrados[df_cadastrados.loc[:,"ID_Cliente"].duplicated(keep = False)].loc[:,"ID_Cliente"]

In [9]:
ids_repetidos

421211    7702516
421268    7602432
421349    7602432
421464    7836971
421698    7213374
           ...   
433158    7282535
433159    7742853
433217    7135270
433666    7091721
433789    7618285
Name: ID_Cliente, Length: 94, dtype: int64

In [10]:
import matplotlib.pyplot as plt

In [11]:
df_cadastrados.iloc[ids_repetidos.index,].sort_values(by = ["ID_Cliente"]).head(20)

ID_Cliente Genero  Tem_Carro  Tem_Casa_Propria  Qtd_Filhos  \
426818     7022197      H          1                 1           3   
425023     7022197      M          0                 1           0   
431545     7022327      M          0                 1           0   
431911     7022327      H          1                 1           0   
425486     7023108      H          1                 1           1   
426488     7023108      M          0                 0           0   
425306     7023651      M          0                 0           0   
421907     7023651      H          1                 0           1   
427778     7024111      H          0                 0           2   
432643     7024111      M          0                 1           0   
424708     7036518      M          0                 1           0   
429078     7036518      M          0                 0           1   
427007     7045794      M          0                 0           0   
428048     7045794      M          0                 1           2   
432271     7045885      M          0                 0           2   
426022     7045885      H          0                 1           0   
432134     7046068      M          0                 1           0   
424618     7046068      M          0                 1           0   
428171     7050948      M          0                 0           0   
430746     7050948      M          0                 1           1   

        Rendimento_Anual   Categoria_de_renda           Grau_Escolaridade  \
426818          135000.0            Empregado                Ensino medio   
425023          450000.0  Associado comercial             Ensino superior   
431545          135000.0  Associado comercial                Ensino medio   
431911          256500.0  Associado comercial             Ensino superior   
425486           67500.0            Empregado                Ensino medio   
426488          135000.0            Empregado                Ensino medio   
425306          225000.0  Associado comercial  Ensino superior incompleto   
421907          157500.0  Associado comercial  Ensino superior incompleto   
427778          157500.0  Associado comercial                Ensino medio   
432643          180000.0            Empregado                Ensino medio   
424708          360000.0            Empregado             Ensino superior   
429078          180000.0            Empregado                Ensino medio   
427007          414000.0            Empregado             Ensino superior   
428048          225000.0            Empregado                Ensino medio   
432271          247500.0  Associado comercial                Ensino medio   
426022           67500.0          Pensionista                Ensino medio   
432134          112500.0            Empregado                Ensino medio   
424618          135000.0          Pensionista                Ensino medio   
428171          135000.0  Associado comercial             Ensino superior   
430746           81000.0            Empregado             Ensino superior   

         Estado_Civil            Moradia  Idade  Tem_Celular  \
426818         Casado       Casa propria   33.0            1   
425023       Separado       Casa propria   54.0            1   
431545       Solteiro       Casa propria   40.0            1   
431911         Casado       Casa propria   59.0            1   
425486         Casado       Casa propria   42.0            1   
426488         Casado       Casa propria   48.0            1   
425306       Solteiro       Casa propria   28.0            1   
421907         Casado       Casa propria   29.0            1   
427778  Uniao estavel       Casa propria   42.0            1   
432643         Casado       Casa propria   60.0            1   
424708         Casado       Casa propria   43.0            1   
429078         Casado       Casa propria   38.0            1   
427007       Separado       Casa propria   57.0            1   
428048         Casado  

In [12]:
df_cadastrados_no_rep = df_cadastrados.drop(ids_repetidos.index) 

In [13]:
print(df_cadastrados_no_rep.loc[:,"ID_Cliente"].value_counts()) # indica que agora o dataframe não tem repetição
print(df_cadastrados_no_rep.loc[:,"ID_Cliente"].value_counts().unique())

5008804    1
6147691    1
6147689    1
6147688    1
6147685    1
          ..
6093492    1
5690720    1
6538938    1
6538937    1
6842885    1
Name: ID_Cliente, Length: 438463, dtype: int64
[1]


In [14]:
df_cadastrados_no_rep.isnull().sum() # não tem dados nulos

ID_Cliente               0
Genero                   0
Tem_Carro                0
Tem_Casa_Propria         0
Qtd_Filhos               0
Rendimento_Anual         0
Categoria_de_renda       0
Grau_Escolaridade        0
Estado_Civil             0
Moradia                  0
Idade                    0
Tem_Celular              0
Tem_telefone_trabalho    0
Tem_telefone_fixo        0
Tem_email                0
Ocupacao                 0
Tamanho_Familia          0
Anos_empregado           0
dtype: int64

In [15]:
df_cadastrados_no_rep.info() # quais são os tipos de dados nas celulas

<class 'pandas.core.frame.DataFrame'>
Int64Index: 438463 entries, 0 to 438556
Data columns (total 18 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   ID_Cliente             438463 non-null  int64  
 1   Genero                 438463 non-null  object 
 2   Tem_Carro              438463 non-null  int64  
 3   Tem_Casa_Propria       438463 non-null  int64  
 4   Qtd_Filhos             438463 non-null  int64  
 5   Rendimento_Anual       438463 non-null  float64
 6   Categoria_de_renda     438463 non-null  object 
 7   Grau_Escolaridade      438463 non-null  object 
 8   Estado_Civil           438463 non-null  object 
 9   Moradia                438463 non-null  object 
 10  Idade                  438463 non-null  float64
 11  Tem_Celular            438463 non-null  int64  
 12  Tem_telefone_trabalho  438463 non-null  int64  
 13  Tem_telefone_fixo      438463 non-null  int64  
 14  Tem_email              438463 non-nu

In [16]:
df_cadastrados_no_rep.describe() # estatística básica

ID_Cliente      Tem_Carro  Tem_Casa_Propria     Qtd_Filhos  \
count  4.384630e+05  438463.000000     438463.000000  438463.000000   
mean   6.021894e+06       0.371906          0.693352       0.427384   
std    5.713554e+05       0.483314          0.461103       0.724879   
min    5.008804e+06       0.000000          0.000000       0.000000   
25%    5.609340e+06       0.000000          0.000000       0.000000   
50%    6.047690e+06       0.000000          1.000000       0.000000   
75%    6.454118e+06       1.000000          1.000000       1.000000   
max    7.999952e+06       1.000000          1.000000      19.000000   

       Rendimento_Anual          Idade  Tem_Celular  Tem_telefone_trabalho  \
count      4.384630e+05  438463.000000     438463.0          438463.000000   
mean       1.875220e+05      43.834032          1.0               0.206129   
std        1.100832e+05      11.472815          0.0               0.404525   
min        2.610000e+04      21.000000          1.0               0.000000   
25%        1.215000e+05      34.000000          1.0               0.000000   
50%        1.611000e+05      43.000000          1.0               0.000000   
75%        2.250000e+05      53.000000          1.0               0.000000   
max        6.750000e+06      69.000000          1.0               1.000000   

       Tem_telefone_fixo      Tem_email  Tamanho_Familia  Anos_empregado  
count      438463.000000  438463.000000    438463.000000   438463.000000  
mean            0.287762       0.108196         2.194461     -165.936036  
std             0.452720       0.310628         0.897194      380.201131  
min             0.000000       0.000000         1.000000    -1000.700000  
25%             0.000000       0.000000         2.000000        1.000000  
50%             0.000000       0.000000         2.000000        4.000000  
75%             1.000000       0.000000         3.000000        8.500000  
max             1.000000       1.000000        20.000000       48.000000

O valor de -1000 "anos empregado" também aparece no conjunto sem os dados repetidos. Provavelmente lidaremos com esse erro de novo mais para frente. Comparando os padrões dos clientes com IDs repetidos, não é possível determinar qual a origem do erro: são identidades completamente diferentes mesmo, pessoas com estados civis diferentes e tamanho familiar diferente (uma hipótese é que fossem casais). A princípio não há como inferir a fonte do erro. A minha hipótese é de que um número de cadastro ficou inválido (a pessoa fechou a conta no banco, por exemplo) e utilizaram o mesmo ID para um novo cliente, sem remover os antigos da base. Dessa forma, não é possível determinar qual o cliente "novo" e o cliente "antigo".

# Desafio 2: Explorar as colunas (features), identificar contínuas, categóricas e binárias

In [17]:
df_cadastrados_no_rep.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 438463 entries, 0 to 438556
Data columns (total 18 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   ID_Cliente             438463 non-null  int64  
 1   Genero                 438463 non-null  object 
 2   Tem_Carro              438463 non-null  int64  
 3   Tem_Casa_Propria       438463 non-null  int64  
 4   Qtd_Filhos             438463 non-null  int64  
 5   Rendimento_Anual       438463 non-null  float64
 6   Categoria_de_renda     438463 non-null  object 
 7   Grau_Escolaridade      438463 non-null  object 
 8   Estado_Civil           438463 non-null  object 
 9   Moradia                438463 non-null  object 
 10  Idade                  438463 non-null  float64
 11  Tem_Celular            438463 non-null  int64  
 12  Tem_telefone_trabalho  438463 non-null  int64  
 13  Tem_telefone_fixo      438463 non-null  int64  
 14  Tem_email              438463 non-nu

In [18]:
df_cadastrados_no_rep.describe()

ID_Cliente      Tem_Carro  Tem_Casa_Propria     Qtd_Filhos  \
count  4.384630e+05  438463.000000     438463.000000  438463.000000   
mean   6.021894e+06       0.371906          0.693352       0.427384   
std    5.713554e+05       0.483314          0.461103       0.724879   
min    5.008804e+06       0.000000          0.000000       0.000000   
25%    5.609340e+06       0.000000          0.000000       0.000000   
50%    6.047690e+06       0.000000          1.000000       0.000000   
75%    6.454118e+06       1.000000          1.000000       1.000000   
max    7.999952e+06       1.000000          1.000000      19.000000   

       Rendimento_Anual          Idade  Tem_Celular  Tem_telefone_trabalho  \
count      4.384630e+05  438463.000000     438463.0          438463.000000   
mean       1.875220e+05      43.834032          1.0               0.206129   
std        1.100832e+05      11.472815          0.0               0.404525   
min        2.610000e+04      21.000000          1.0               0.000000   
25%        1.215000e+05      34.000000          1.0               0.000000   
50%        1.611000e+05      43.000000          1.0               0.000000   
75%        2.250000e+05      53.000000          1.0               0.000000   
max        6.750000e+06      69.000000          1.0               1.000000   

       Tem_telefone_fixo      Tem_email  Tamanho_Familia  Anos_empregado  
count      438463.000000  438463.000000    438463.000000   438463.000000  
mean            0.287762       0.108196         2.194461     -165.936036  
std             0.452720       0.310628         0.897194      380.201131  
min             0.000000       0.000000         1.000000    -1000.700000  
25%             0.000000       0.000000         2.000000        1.000000  
50%             0.000000       0.000000         2.000000        4.000000  
75%             1.000000       0.000000         3.000000        8.500000  
max             1.000000       1.000000        20.000000       48.000000

In [19]:
df_cadastrados_no_rep.head(2)

ID_Cliente Genero  Tem_Carro  Tem_Casa_Propria  Qtd_Filhos  \
0     5008804      H          1                 1           0   
1     5008805      H          1                 1           0   

   Rendimento_Anual Categoria_de_renda Grau_Escolaridade   Estado_Civil  \
0          427500.0          Empregado   Ensino superior  Uniao estavel   
1          427500.0          Empregado   Ensino superior  Uniao estavel   

   Moradia  Idade  Tem_Celular  Tem_telefone_trabalho  Tem_telefone_fixo  \
0  Alugado   33.0            1                      1                  0   
1  Alugado   33.0            1                      1                  0   

   Tem_email Ocupacao  Tamanho_Familia  Anos_empregado  
0          0   Outros              2.0            12.4  
1          0   Outros              2.0            12.4

Features contínuas: "Rendimento_Anual", "Idade", "Tamanho Família", "Anos empregado", "Quantidade de filhos".

Features categóricas: "Categoria de renda", "Grau escolaridade", "Estado civil", "Moradia", "Ocupação".

Features categóricas binárias: "Genero", "Tem carro", "Tem casa propria", "Tem celular", "Tem telefone trabalho", "Tem telefone fixo", "Tem email".

In [20]:
cont = ["Rendimento_Anual", "Idade", "Tamanho_Familia", "Anos_empregado", "Qtd_filhos"]
cat = ["Categoria_de_renda", "Grau_Escolaridade", "Estado_Civil", "Moradia", "Ocupacao"]
bin = ["Genero", "Tem_Carro", "Tem_Casa_Propria", "Tem_Celular", "Tem_telefone_trabalho", "Tem_telefone_fixo", "Tem_email"]